py/dml/messages.py

```diff
@@ -2110,16 +2110,18 @@ def log(self):
             self.other_site,
             "original non-throwing declaration")
 
-class WNEGCONSTCOMP(DMLWarning):
-    """DML uses a special method when comparing an unsigned and signed integer,
-    meaning that comparing a negative constant to an unsigned integer always
-    has the same result, which is usually not the intended behaviour."""
+class WTYPELIMITS(DMLWarning):
+    """When comparing an integer with a constant outside the range of the
+    integer's type, the comparison always has the same result. One
+    common example is that a value of type `uint64` always is
+    considered to be different from -1; in this case, the constant
+    can be rewritten as `cast(-1, uint64)`.
+    """
     def __init__(self, site, expr, ty):
-        DMLWarning.__init__(self, site)
+        super().__init__(site)
         self.expr = expr
         self.ty = ty
-    fmt = ("Comparing negative constant to unsigned integer has a constant "
-           + "result")
+    fmt = 'Comparison with constant outside the range of data type'
```

Comment:
I'm slightly worried this will be hard to understand... maybe
> comparing a value of type '%s' against a constant outside the range of that type

possibly s/against/with. This might also need some tweaking in the case of e.g. `cast(a, uint32) > 0xffffffff`, where  `0xffffffff` is in _range_ of uint32, and yet this check is still always constant false. (Your code does handle this case).

Reply:
I changed to gcc's phrasing, "comparison is always false/true due to limited range of data type", at the cost of a flag programming hack to know whether we are in the `!=` inverted case